listenbrainz/bigquery/__init__.py

```diff
@@ -0,0 +1,207 @@
+
+    if not APP_CREDENTIALS_FILE:
+        logger.error("The GOOGLE_APPLICATIONS_CREDENTIALS variable is undefined, cannot connect to BigQuery")
+        raise NoCredentialsVariableException
+
+    if not os.path.exists(APP_CREDENTIALS_FILE):
+        logger.error("The BigQuery credentials file does not exist, cannot connect to BigQuery")
+        raise NoCredentialsFileException
+
+    credentials = GoogleCredentials.get_application_default()
+    return discovery.build('bigquery', 'v2', credentials=credentials)
+
+
+def get_parameters_dict(parameters):
+    """ Converts a list of parameters to be passed to BigQuery into the standard format that the API requires.
+        The format can be seen here:
+        https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/python/latest/bigquery_v2.jobs.html#query
+
+        Args: parameters: a list of dictionaries of the following form
+                {
+                    "name" (str): name of the parameter,
+                    "type" (str): type of the parameter,
+                    "value" (str): value of the parameter
+                }
+
+        Returns: A list of dictionaries that can be passed to the API call
+    """
+
+    if not parameters:
+        return None
+
+    bq_params = []
+    for param in parameters:
+        # construct parameter dict
+        temp = {}
+        temp["name"] = param["name"]
+        temp["parameterType"] = {
+            "type": param["type"],
+        }
+        temp["parameterValue"] = {
+            "value": param["value"],
+        }
+
+        # append parameter dict to main list
+        bq_params.append(temp)
+
+    return bq_params
+
+
+def wait_for_completion(projectId, jobId):
+    """ Make requests periodically until the passed job has been completed """
+
+    while True:
+
+        try:
+            job = bigquery.jobs().get(projectId=projectId, jobId=jobId).execute(num_retries=5)
+        except googleapiclient.errors.HttpError as err:
+            logger.error("HttpError while waiting for completion of job: {}".format(err))
+            time.sleep(JOB_COMPLETION_CHECK_DELAY)
+            continue
+
+        if job["status"]["state"] == "DONE":
+            return
+        else:
+            time.sleep(JOB_COMPLETION_CHECK_DELAY)
+
+
+def format_results(data):
+    """ The data returned by BigQuery contains a dict for the schema and a seperate dict for
+        the rows. This function formats the data into a form similar to the data returned
+        by sqlalchemy i.e. a dictionary keyed by row names
+    """
+
+    formatted_data = []
+    for row in data['rows']:
+        formatted_row = {}
+        for index, val in enumerate(row['f']):
+            formatted_row[data['schema']['fields'][index]['name']] = val['v']
+        formatted_data.append(formatted_row)
+    return formatted_data
+
+
+def run_query(bigquery, query, parameters=None, dml=False):
```

Comment:
Also this dual purpose function gives me the willies... But I am not sure if it is worth cleaning up...

Reply:
Data Manipulation Language (https://cloud.google.com/bigquery/docs/reference/standard-sql/data-manipulation-language)

I can break it into two functions if that's the better case for future. I initially decided to go with this because I thought `run_query` should run all types of queries.

Reply:
I think it is fine for now -- if we need to go back and change anything, then let's revisit this.